newrelic/metric.go

```diff
@@ -0,0 +1,226 @@
+	"google.golang.org/appengine/log"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+// MetricConfig defines configuration file parameters for a specific metric imported from New Relic.
+type MetricConfig struct {
+	APIKey        string `yaml:"api_key" validate:"nonzero"`
+	ApplicationId string `yaml:"application_id" validate:"nonzero"`
+	MetricName    string `validate:"nonzero" yaml:"metric_name"`
+	MetricValue   string `validate:"nonzero" yaml:"metric_value"`
+	Cumulative    bool
+}
+
+// Metric defines a New Relic-based metric. It implements the SourceMetric interface.
+type Metric struct {
+	Name   string
+	config *MetricConfig
+	client *Client
+}
+
+type timeslice struct {
+	From   string             `json:"from"`
+	To     string             `json:"to"`
+	Values map[string]float64 `json:"values"`
+}
+type data struct {
+	Name       string      `json:"name"`
+	Timeslices []timeslice `json:"timeslices"`
+}
+type metricData struct {
+	From            string   `json:"from"`
+	To              string   `json:"to"`
+	MetricsNotFound []string `json:"metrics_not_found,omitempty"`
+	MetricsFound    []string `json:"metrics_found,omitempty"`
+	Metrics         []data   `json:"metrics`
+}
+type responseData struct {
+	MetricData metricData `json:"metric_data"`
+}
+
+// Client defines the information necessary to query New Relic
+type Client struct {
+	APIKey        string
+	ApplicationId string
+	BaseUrl       string
+}
+
+// Query returns a unique name for the query
+func (m *Metric) Query() string {
+	return fmt.Sprintf("%v:%v", m.config.MetricName, m.config.MetricValue)
+}
+
+// QueryMetrics returns a metric descriptor and a slice of TimeSeries for the metric.
+// all elements of the TimeSeries must be the same metric descriptor and contain a single point
+func (m *Metric) QueryMetrics(ctx context.Context, start time.Time, end time.Time, name string, value string) (*metricpb.MetricDescriptor, []*monitoringpb.TimeSeries, error) {
+	c := m.client
+
+	var ret []*monitoringpb.TimeSeries
+	tmpStart := start
+
+	for continueLoop := true; continueLoop; {
+		// Always use 3 hour windows, anything more and points will get aggregated to greater than 1 minute granularity
+		tmpEnd := tmpStart.Add(3 * time.Hour)
+		if tmpEnd.After(end) {
+			tmpEnd = end
+			continueLoop = false
+		}
+
+		// TODO: add multi-page support
+		url := fmt.Sprintf("%s/v2/applications/%v/metrics/data.json",
+			c.BaseUrl, c.ApplicationId)
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		q := req.URL.Query()
+		q.Add("names[]", name)
+		q.Add("values[]", value)
+		q.Add("from", tmpStart.Format(time.RFC3339))
+		q.Add("to", tmpEnd.Format(time.RFC3339))
+
+		req.URL.RawQuery = q.Encode()
+		req.Header.Set("X-Api-Key", c.APIKey)
+		httpClient := &http.Client{}
+		resp, err := httpClient.Do(req)
```

Comment:
Do you need a custom client? If not, I would suggest to just use `http.DefaultClient.Do(req)`.

Reply:
This comment does not seem to be addressed.